src/storage/file_storage.rs

```diff
@@ -0,0 +1,176 @@
+}
+
+impl FileStorage {
+    /// Create a new file storage with a diagnostic label and file path
+    pub fn new(label: String, path: &Path) -> Self {
+        Self {
+            label: label.clone(),
+            log_prefix: format!("{} {}", LOG_PREFIX, label).to_string(),
+            path: PathBuf::from(path),
+            timestamp: SystemTime::now(),
+        }
+    }
+
+    /// Read data from disk
+    ///
+    /// Data is read as a key value pairs separated by a symbol and stored
+    /// in a [HashMap] with [ResourceId] key and a generic V value. A handler
+    /// is called on the data after reading it.
+    pub fn read_from_disk<V>(
+        &self,
+        handle: impl FnOnce(HashMap<ResourceId, V>),
+    ) -> Result<()>
+    where
+        V: FromStr,
+        ArklibError: From<<V as FromStr>::Err>,
+    {
+        let new_timestamp = fs::metadata(&self.path)?.modified()?;
+        log::info!(
+            "timestamp of storage file {:?} is {:?}",
+            self.path,
+            self.timestamp
+        );
+
+        if self.timestamp >= new_timestamp {
+            return Ok(());
+        }
+
+        log::info!("the file was modified externally, merging");
+
+        let file = fs::File::open(&self.path)?;
+        let reader = BufReader::new(file);
+        let mut lines = reader.lines();
+
+        match lines.next() {
+            Some(Ok(header)) => {
+                self.verify_version(&header)?;
+
+                let mut value_by_id = HashMap::new();
+                for line in lines {
+                    let line = line?;
+                    if line.is_empty() {
+                        continue;
+                    }
+
+                    let parts: Vec<&str> =
+                        line.split(KEY_VALUE_SEPARATOR).collect();
+                    let id = ResourceId::from_str(parts[0])?;
+                    let value = V::from_str(parts[1])?;
+
+                    // TODO: neutral value check when introducing monoids
+
+                    value_by_id.insert(id, value);
+                }
+
+                if !value_by_id.is_empty() {
+                    handle(value_by_id);
+                }
+            }
+            Some(Err(e)) => return Err(e.into()),
+            None => (),
+        }
+
+        Ok(())
+    }
+
+    /// Write data to file
+    ///
+    /// Data is a key-value mapping between [ResourceId] and a generic Value
+    pub fn write_to_disk<V>(
+        &mut self,
+        value_by_id: &HashMap<ResourceId, V>,
+    ) -> Result<()>
+    where
+        V: Display,
+    {
+        fs::create_dir_all(self.path.parent().unwrap())?;
+        let file = File::create(&self.path)?;
+        let mut writer = BufWriter::new(file);
+
+        writer.write_all(
+            format!("{}{}\n", STORAGE_VERSION_PREFIX, STORAGE_VERSION)
+                .as_bytes(),
+        )?;
+
+        for (id, value) in value_by_id {
+            // TODO: Is neutral check might be needed when monoids are introduced
```

Comment:
TODOs should be resolved before merging.

Reply:
Same as previous comment about `isNeutral` function